Reviving parallel exception code for PETSc 3.6

jwpeterson opened the pull request:
This branch revives @friedmud's exception throwing code and expands on it a bit.  It works for me on my laptop using PETSc-3.6, I'm putting it up here to see how the build system's PETSc deals with it.  Eventually the test may need to be restricted to `petsc_version = '>=3.6.0'`, although that version does not exist yet...

Refs #3777

friedmud committed:
Parallel exception handling.

The basic idea is this:

1.  Exceptions are thrown _and_ caught within threaded regions (an exception never traverses a threaded call).
2.  When it's caught within the threaded region a flag is set in MOOSE.
3.  Other threads can check that flag to see if they should quit early (or they can keep working - who cares).
4.  Once the threaded region is finished the flag is then checked and communicated globally over MPI.
5.  If there was an exception PETSc is informed and the rest of the operation (like residual computation) is (possibly) aborted.
6.  PETSc should stop the current solve (it's currently not)
6.  The Executioner should get the flag back from PETSc that the solve failed and do what it normally does when the solve fails (like cut timesteps).

refs #3777

Rebased from Aug 2014 MOOSE to June 2015 MOOSE by John W. Peterson <jwpeterson@gmail.com>.

Conflicts:
framework/include/base/ThreadedElementLoopBase.h
framework/src/base/AuxiliarySystem.C
framework/src/base/FEProblem.C
framework/src/base/NonlinearSystem.C
test/src/executioners/ExceptionSteady.C
test/tests/misc/exception/parallel_exception_residual.i

jwpeterson committed:
Misc. commits while working on Exception:

* It is no longer necessary to set type=Exception in the exception
  tests, the Exception is thrown because of the ExceptionKernel not
  the Excecutioner.

* FEProblem::setException() should take a const reference.

* We don't want to call checkExceptionAndStopSolve() after computing the
  initial residual because it throws an exception.  We need to trust
  that the exception has already been handled at a lower level and PETSc
  will return a diverged flag during it's next reduction.

* Throw on call number 1000 in ExceptionKernel.

* parallel_exception_residual.i doesn't need -pc_factor_shift_type if
  we put the Kernels on all blocks.

* More specific message from ExceptionKernel::computeQpResidual()

* Catch and don't re-throw in NonlinearSystem::computeJacobian()

* Modify the ExceptionKernel to only throw after a specific number of
  calls to computeQpJacobian() so we don't throw during the routine
  where the sparsity of the preconditioner is constructed.

* Throw an exception from ExceptionKernel rather than calling stopSolve().

* Set NaN in the residual vector during stopSolve().

* Adding transient versions of tests which throw from the residual and Jacobian.

* Once checkExceptionAndStopSolve() has handled the exception, set
  _has_exception to false.

* In the ExceptionKernel, increment the call number before we throw,
  so the Kernel can continue to be used after that point instead of
  throwing exceptions every time it is called subsequently.

* Remove the original exception_test, which was already deleted.

* Removing original exception residual and jacobian tests.
  They are redundant with the transient version of these tests.

* Changing the parallel_exception_initial_condition test to RunApp
  We are testing that we get the output string:

  'MooseException thrown during initial condition computation'

  This test is somewhat independent of the the residual test, since the
  exception is thrown from a different invocation of computeResidual().

Refs #3777

moosebuild commented:
Results of testing 60f641bfcc7a73eac8632767003b38d1c1d564c6 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17255

moosebuild commented:
Results of testing 60f641bfcc7a73eac8632767003b38d1c1d564c6 using moose_PR_test_dbg recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17257

moosebuild commented:
Results of testing 60f641bfcc7a73eac8632767003b38d1c1d564c6 using moose_PR_test recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17256

moosebuild commented:
Results of testing 60f641bfcc7a73eac8632767003b38d1c1d564c6 using moose_PR_app_tests recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17258

jwpeterson committed:
Add throw() specification to MooseException destructor.

This declares that it won't throw an exception.

Refs #3777.

moosebuild commented:
Results of testing 98d02e1e9708b89b245f9a7bb37d5ab80ce4777d using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17259

moosebuild commented:
Results of testing 98d02e1e9708b89b245f9a7bb37d5ab80ce4777d using moose_PR_app_tests recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17262

moosebuild commented:
Results of testing 98d02e1e9708b89b245f9a7bb37d5ab80ce4777d using moose_PR_test recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17260

moosebuild commented:
Results of testing 98d02e1e9708b89b245f9a7bb37d5ab80ce4777d using moose_PR_test_dbg recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17261

jwpeterson committed:
Restrict new test to only run on PETSc >= 3.6.0

Otherwise, the error message is:

```
Time Step  0, time = 0
                dt = 0

Time Step  1, time = 0.01
                dt = 0.01
 0 Nonlinear |R| = 2.449490e+00
      0 Linear |R| = 2.449490e+00
      1 Linear |R| = 1.133359e-02
MooseException thrown during residual calculation
      2 Linear |R| = nan
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Petsc has generated inconsistent data
[0]PETSC ERROR: Computed Nan differencing parameter h
[0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.5.2, Sep, 08, 2014
```

moosebuild commented:
Results of testing 675b730bb514b2fa6e00ad5980b4846e83bd9f1a using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17265

moosebuild commented:
Results of testing 675b730bb514b2fa6e00ad5980b4846e83bd9f1a using moose_PR_app_tests recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17268

jwpeterson committed:
Fix unused variable warning.

moosebuild commented:
Results of testing 35d87634ff476ee93cfdeccd02a4509a990869ba using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17269

moosebuild commented:
Results of testing 675b730bb514b2fa6e00ad5980b4846e83bd9f1a using moose_PR_test recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17266

moosebuild commented:
Results of testing 35d87634ff476ee93cfdeccd02a4509a990869ba using moose_PR_app_tests recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17272

moosebuild commented:
Results of testing 35d87634ff476ee93cfdeccd02a4509a990869ba using moose_PR_test recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17270

jwpeterson commented:
This branch requires a submodule update in BISON and doesn't seem to work quite right in parallel... closing for the moment.

moosebuild commented:
Results of testing 675b730bb514b2fa6e00ad5980b4846e83bd9f1a using moose_PR_test_dbg recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17267

moosebuild commented:
Results of testing 35d87634ff476ee93cfdeccd02a4509a990869ba using moose_PR_test_dbg recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17271
